ParametersLatex.rst
.. Dynamically generated list of documented parameters
.. This page was generated using Tools\/autotest\/param\_metadata\/param\_parse\.py

.. DO NOT EDIT


.. _parameters:

Complete Parameter List
=======================

This is a complete list of the parameters which can be set \(e\.g\. via the MAVLink protocol\) to control vehicle behaviour\. They are stored in persistent storage on the vehicle\.

This list is automatically generated from the latest ardupilot source code\, and so may contain parameters which are not yet in the stable released versions of the code\.




.. _parameters_ArduCopter:

ArduCopter Parameters
---------------------


.. _FORMAT_VERSION:

FORMAT\_VERSION: Eeprom format version number
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

| *Note: This parameter is for advanced users*

This value is incremented when changes are made to the eeprom format


.. _SYSID_THISMAV:

SYSID\_THISMAV: MAVLink system ID of this vehicle
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

| *Note: This parameter is for advanced users*

Allows setting an individual MAVLink system id for this vehicle to distinguish it from others on the same network


+----------+
| Range    |
+==========+
| 1 to 255 |
+----------+




.. _SYSID_MYGCS:

SYSID\_MYGCS: My ground station number
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

| *Note: This parameter is for advanced users*

Allows restricting radio overrides to only come from my ground station


+-----------+----------+
| Increment | Range    |
+===========+==========+
| 1         | 1 to 255 |
+-----------+----------+




.. _PILOT_THR_FILT:

PILOT\_THR\_FILT: Throttle filter cutoff
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

| *Note: This parameter is for advanced users*

Throttle filter cutoff \(Hz\) \- active whenever altitude control is inactive \- 0 to disable


+-----------+---------+-------+
| Increment | Range   | Units |
+===========+=========+=======+
| .5        | 0 to 10 | hertz |
+-----------+---------+-------+




.. _PILOT_TKOFF_ALT:

PILOT\_TKOFF\_ALT: Pilot takeoff altitude
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~


Altitude that altitude control modes will climb to when a takeoff is triggered with the throttle stick\.


+-----------+---------------+-------------+
| Increment | Range         | Units       |
+===========+===============+=============+
| 10        | 0.0 to 1000.0 | centimeters |
+-----------+---------------+-------------+




.. _PILOT_THR_BHV:

PILOT\_THR\_BHV: Throttle stick behavior
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~


Bitmask containing various throttle stick options\. TX with sprung throttle can set PILOT\_THR\_BHV to \"1\" so motor feedback when landed starts from mid\-stick instead of bottom of stick\.


+-----+-------------------------------+
| Bit | Meaning                       |
+=====+===============================+
| 0   | Feedback from mid stick       |
+-----+-------------------------------+
| 1   | High throttle cancels landing |
+-----+-------------------------------+
| 2   | Disarm on land detection      |
+-----+-------------------------------+




.. _TELEM_DELAY:

TELEM\_DELAY: Telemetry startup delay
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

| *Note: This parameter is for advanced users*

The amount of time \(in seconds\) to delay radio telemetry to prevent an Xbee bricking on power up


+-----------+---------+---------+
| Increment | Range   | Units   |
+===========+=========+=========+
| 1         | 0 to 30 | seconds |
+-----------+---------+---------+




.. _GCS_PID_MASK:

GCS\_PID\_MASK: GCS PID tuning mask
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

| *Note: This parameter is for advanced users*

bitmask of PIDs to send MAVLink PID\_TUNING messages for


+-----+---------+
| Bit | Meaning |
+=====+=========+
| 0   | Roll    |
+-----+---------+
| 1   | Pitch   |
+-----+---------+
| 2   | Yaw     |
+-----+---------+
| 3   | AccelZ  |
+-----+---------+




.. _RTL_ALT:

RTL\_ALT: RTL Altitude
~~~~~~~~~~~~~~~~~~~~~~


The minimum alt above home the vehicle will climb to before returning\.  If the vehicle is flying higher than this value it will return at its current altitude\.


+-----------+--------------+-------------+
| Increment | Range        | Units       |
+===========+==============+=============+
| 1         | 30 to 300000 | centimeters |
+-----------+--------------+-------------+




.. _RTL_CONE_SLOPE:

RTL\_CONE\_SLOPE: RTL cone slope
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~


Defines a cone above home which determines maximum climb


+-------+----------+
| Value | Meaning  |
+=======+==========+
| 0     | Disabled |
+-------+----------+
| 1     | Shallow  |
+-------+----------+
| 3     | Steep    |
+-------+----------+




+-----------+-------------+
| Increment | Range       |
+===========+=============+
| .1        | 0.5 to 10.0 |
+-----------+-------------+




.. _RTL_SPEED:

RTL\_SPEED: RTL speed
~~~~~~~~~~~~~~~~~~~~~


Defines the speed in cm\/s which the aircraft will attempt to maintain horizontally while flying home\. If this is set to zero\, WPNAV\_SPEED will be used instead\.


+-----------+-----------+------------------------+
| Increment | Range     | Units                  |
+===========+===========+========================+
| 50        | 0 to 2000 | centimeters per second |
+-----------+-----------+------------------------+




.. _RTL_ALT_FINAL:

RTL\_ALT\_FINAL: RTL Final Altitude
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~


This is the altitude the vehicle will move to as the final stage of Returning to Launch or after completing a mission\.  Set to zero to land\.


+-----------+-----------+-------------+
| Increment | Range     | Units       |
+===========+===========+=============+
| 1         | 0 to 1000 | centimeters |
+-----------+-----------+-------------+




.. _RTL_CLIMB_MIN:

RTL\_CLIMB\_MIN: RTL minimum climb
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~


The vehicle will climb this many cm during the initial climb portion of the RTL


+-----------+-----------+-------------+
| Increment | Range     | Units       |
+===========+===========+=============+
| 10        | 0 to 3000 | centimeters |
+-----------+-----------+-------------+




.. _RTL_LOIT_TIME:

RTL\_LOIT\_TIME: RTL loiter time
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~


Time \(in milliseconds\) to loiter above home before beginning final descent


+-----------+------------+--------------+
| Increment | Range      | Units        |
+===========+============+==============+
| 1000      | 0 to 60000 | milliseconds |
+-----------+------------+--------------+




.. _RTL_ALT_TYPE:

RTL\_ALT\_TYPE: RTL mode altitude type
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~


RTL altitude type\.  Set to 1 for Terrain following during RTL and then set WPNAV\_RFND\_USE\=1 to use rangefinder or WPNAV\_RFND\_USE\=0 to use Terrain database


+-------+------------------+
| Value | Meaning          |
+=======+==================+
| 0     | Relative to Home |
+-------+------------------+
| 1     | Terrain          |
+-------+------------------+




.. _FS_GCS_ENABLE:

FS\_GCS\_ENABLE: Ground Station Failsafe Enable
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~


Controls whether failsafe will be invoked \(and what action to take\) when connection with Ground station is lost for at least 5 seconds\. See FS\_OPTIONS param for additional actions\, or for cases allowing Mission continuation\, when GCS failsafe is enabled\.


+-------+----------------------------------------------------------------------------+
| Value | Meaning                                                                    |
+=======+============================================================================+
| 0     | Disabled/NoAction                                                          |
+-------+----------------------------------------------------------------------------+
| 1     | RTL                                                                        |
+-------+----------------------------------------------------------------------------+
| 2     | RTL or Continue with Mission in Auto Mode (Removed in 4.0+-see FS_OPTIONS) |
+-------+----------------------------------------------------------------------------+
| 3     | SmartRTL or RTL                                                            |
+-------+----------------------------------------------------------------------------+
| 4     | SmartRTL or Land                                                           |
+-------+----------------------------------------------------------------------------+
| 5     | Land                                                                       |
+-------+----------------------------------------------------------------------------+
| 6     | Auto DO_LAND_START or RTL                                                  |
+-------+----------------------------------------------------------------------------+
| 7     | Brake or Land                                                              |
+-------+----------------------------------------------------------------------------+




.. _GPS_HDOP_GOOD:

GPS\_HDOP\_GOOD: GPS Hdop Good
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

| *Note: This parameter is for advanced users*

GPS Hdop value at or below this value represent a good position\.  Used for pre\-arm checks


+------------+
| Range      |
+============+
| 100 to 900 |
+------------+




.. _SUPER_SIMPLE:

SUPER\_SIMPLE: Super Simple Mode
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~


Bitmask to enable Super Simple mode for some flight modes\. Setting this to Disabled\(0\) will disable Super Simple Mode\. The bitmask is for flight mode switch positions


+-----+------------+
| Bit | Meaning    |
+=====+============+
| 0   | SwitchPos1 |
+-----+------------+
| 1   | SwitchPos2 |
+-----+------------+
| 2   | SwitchPos3 |
+-----+------------+
| 3   | SwitchPos4 |
+-----+------------+
| 4   | SwitchPos5 |
+-----+------------+
| 5   | SwitchPos6 |
+-----+------------+




.. _WP_YAW_BEHAVIOR:

WP\_YAW\_BEHAVIOR: Yaw behaviour during missions
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~


Determines how the autopilot controls the yaw during missions and RTL


+-------+-------------------------------+
| Value | Meaning                       |
+=======+===============================+
| 0     | Never change yaw              |
+-------+-------------------------------+
| 1     | Face next waypoint            |
+-------+-------------------------------+
| 2     | Face next waypoint except RTL |
+-------+-------------------------------+
| 3     | Face along GPS course         |
+-------+-------------------------------+




.. _LAND_SPEED:

LAND\_SPEED: Land speed
~~~~~~~~~~~~~~~~~~~~~~~


The descent speed for the final stage of landing in cm\/s


+-----------+-----------+------------------------+
| Increment | Range     | Units                  |
+===========+===========+========================+
| 10        | 30 to 200 | centimeters per second |
+-----------+-----------+------------------------+




.. _LAND_SPEED_HIGH:

LAND\_SPEED\_HIGH: Land speed high
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~


The descent speed for the first stage of landing in cm\/s\. If this is zero then WPNAV\_SPEED\_DN is used


+-----------+----------+------------------------+
| Increment | Range    | Units                  |
+===========+==========+========================+
| 10        | 0 to 500 | centimeters per second |
+-----------+----------+------------------------+




.. _PILOT_SPEED_UP:

PILOT\_SPEED\_UP: Pilot maximum vertical speed ascending
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~


The maximum vertical ascending velocity the pilot may request in cm\/s


+-----------+-----------+------------------------+
| Increment | Range     | Units                  |
+===========+===========+========================+
| 10        | 50 to 500 | centimeters per second |
+-----------+-----------+------------------------+




.. _PILOT_ACCEL_Z:

PILOT\_ACCEL\_Z: Pilot vertical acceleration

ArduCopter/config.h
//
#pragma once

//////////////////////////////////////////////////////////////////////////////
//////////////////////////////////////////////////////////////////////////////
//
// WARNING WARNING WARNING WARNING WARNING WARNING WARNING WARNING WARNING
//
//  DO NOT EDIT this file to adjust your configuration.  Create your own
//  APM_Config.h and use APM_Config.h.example as a reference.
//
// WARNING WARNING WARNING WARNING WARNING WARNING WARNING WARNING WARNING
///
//////////////////////////////////////////////////////////////////////////////
//////////////////////////////////////////////////////////////////////////////
//
// Default and automatic configuration details.
//
// Notes for maintainers:
//
// - Try to keep this file organised in the same order as APM_Config.h.example
//
#include "defines.h"

///
/// DO NOT EDIT THIS INCLUDE - if you want to make a local change, make that
/// change in your local copy of APM_Config.h.
///
#include "APM_Config.h"
#include <AP_ADSB/AP_ADSB_config.h>
#include <AP_Follow/AP_Follow_config.h>
#include <AC_Avoidance/AC_Avoidance_config.h>

//////////////////////////////////////////////////////////////////////////////
//////////////////////////////////////////////////////////////////////////////
// HARDWARE CONFIGURATION AND CONNECTIONS
//////////////////////////////////////////////////////////////////////////////
//////////////////////////////////////////////////////////////////////////////

#ifndef CONFIG_HAL_BOARD
#error CONFIG_HAL_BOARD must be defined to build ArduCopter
#endif

#ifndef ARMING_DELAY_SEC
    # define ARMING_DELAY_SEC 2.0f
#endif

//////////////////////////////////////////////////////////////////////////////
// FRAME_CONFIG
//
#ifndef FRAME_CONFIG
 # define FRAME_CONFIG   MULTICOPTER_FRAME
#endif

/////////////////////////////////////////////////////////////////////////////////
// TradHeli defaults
#if FRAME_CONFIG == HELI_FRAME
  # define RC_FAST_SPEED                        125
  # define WP_YAW_BEHAVIOR_DEFAULT              WP_YAW_BEHAVIOR_LOOK_AHEAD
#endif

//////////////////////////////////////////////////////////////////////////////
// PWM control
// default RC speed in Hz
#ifndef RC_FAST_SPEED
   #   define RC_FAST_SPEED 490
#endif

//////////////////////////////////////////////////////////////////////////////
// Rangefinder
//

#ifndef RANGEFINDER_FILT_DEFAULT
 # define RANGEFINDER_FILT_DEFAULT 0.5f     // filter for rangefinder distance
#endif

#ifndef SURFACE_TRACKING_TIMEOUT_MS
 # define SURFACE_TRACKING_TIMEOUT_MS  1000 // surface tracking target alt will reset to current rangefinder alt after this many milliseconds without a good rangefinder alt
#endif

#ifndef MAV_SYSTEM_ID
 # define MAV_SYSTEM_ID          1
#endif

// prearm GPS hdop check
#ifndef GPS_HDOP_GOOD_DEFAULT
 # define GPS_HDOP_GOOD_DEFAULT         140     // minimum hdop that represents a good position.  used during pre-arm checks if fence is enabled
#endif

// missing terrain data failsafe
#ifndef FS_TERRAIN_TIMEOUT_MS
 #define FS_TERRAIN_TIMEOUT_MS          5000     // 5 seconds of missing terrain data will trigger failsafe (RTL)
#endif

// pre-arm baro vs inertial nav max alt disparity
#ifndef PREARM_MAX_ALT_DISPARITY_CM
 # define PREARM_MAX_ALT_DISPARITY_CM       100     // barometer and inertial nav altitude must be within this many centimeters
#endif

//////////////////////////////////////////////////////////////////////////////
//  EKF Failsafe
#ifndef FS_EKF_ACTION_DEFAULT
 # define FS_EKF_ACTION_DEFAULT         FS_EKF_ACTION_LAND  // EKF failsafe triggers land by default
#endif
#ifndef FS_EKF_THRESHOLD_DEFAULT
 # define FS_EKF_THRESHOLD_DEFAULT      0.8f    // EKF failsafe's default compass and velocity variance threshold above which the EKF failsafe will be triggered
#endif

#ifndef EKF_ORIGIN_MAX_ALT_KM
 # define EKF_ORIGIN_MAX_ALT_KM         50   // EKF origin and home must be within 50km vertically
#endif

#ifndef FS_EKF_FILT_DEFAULT
# define FS_EKF_FILT_DEFAULT     5.0f    // frequency cutoff of EKF variance filters
#endif

//////////////////////////////////////////////////////////////////////////////
//  Auto Tuning
#ifndef AUTOTUNE_ENABLED
 # define AUTOTUNE_ENABLED  1
#endif

//////////////////////////////////////////////////////////////////////////////
// Nav-Guided - allows external nav computer to control vehicle
#ifndef AC_NAV_GUIDED
 # define AC_NAV_GUIDED    1
#endif

//////////////////////////////////////////////////////////////////////////////
// Acro - fly vehicle in acrobatic mode
#ifndef MODE_ACRO_ENABLED
# define MODE_ACRO_ENABLED 1
#endif

//////////////////////////////////////////////////////////////////////////////
// Auto mode - allows vehicle to trace waypoints and perform automated actions
#ifndef MODE_AUTO_ENABLED
# define MODE_AUTO_ENABLED 1
#endif

//////////////////////////////////////////////////////////////////////////////
// Brake mode - bring vehicle to stop
#ifndef MODE_BRAKE_ENABLED
# define MODE_BRAKE_ENABLED 1
#endif

//////////////////////////////////////////////////////////////////////////////
// Circle - fly vehicle around a central point
#ifndef MODE_CIRCLE_ENABLED
# define MODE_CIRCLE_ENABLED 1
#endif

//////////////////////////////////////////////////////////////////////////////
// Drift - fly vehicle in altitude-held, coordinated-turn mode
#ifndef MODE_DRIFT_ENABLED
# define MODE_DRIFT_ENABLED 1
#endif

//////////////////////////////////////////////////////////////////////////////
// flip - fly vehicle in flip in pitch and roll direction mode
#ifndef MODE_FLIP_ENABLED
# define MODE_FLIP_ENABLED 1
#endif

//////////////////////////////////////////////////////////////////////////////
// Follow - follow another vehicle or GCS
#ifndef MODE_FOLLOW_ENABLED
#if AP_FOLLOW_ENABLED && AP_AVOIDANCE_ENABLED
#define MODE_FOLLOW_ENABLED 1
#else
#define MODE_FOLLOW_ENABLED 0
#endif
#endif

//////////////////////////////////////////////////////////////////////////////
// Guided mode - control vehicle's position or angles from GCS
#ifndef MODE_GUIDED_ENABLED
# define MODE_GUIDED_ENABLED 1
#endif

//////////////////////////////////////////////////////////////////////////////
// GuidedNoGPS mode - control vehicle's angles from GCS
#ifndef MODE_GUIDED_NOGPS_ENABLED
# define MODE_GUIDED_NOGPS_ENABLED 1
#endif

//////////////////////////////////////////////////////////////////////////////
// Loiter mode - allows vehicle to hold global position
#ifndef MODE_LOITER_ENABLED
# define MODE_LOITER_ENABLED 1
#endif

//////////////////////////////////////////////////////////////////////////////
// Position Hold - enable holding of global position
#ifndef MODE_POSHOLD_ENABLED
# define MODE_POSHOLD_ENABLED 1
#endif

//////////////////////////////////////////////////////////////////////////////
// RTL - Return To Launch
#ifndef MODE_RTL_ENABLED
# define MODE_RTL_ENABLED 1
#endif

//////////////////////////////////////////////////////////////////////////////
// SmartRTL - allows vehicle to retrace a (loop-eliminated) breadcrumb home
#ifndef MODE_SMARTRTL_ENABLED
# define MODE_SMARTRTL_ENABLED 1
#endif

//////////////////////////////////////////////////////////////////////////////
// Sport - fly vehicle in rate-controlled (earth-frame) mode
#ifndef MODE_SPORT_ENABLED
# define MODE_SPORT_ENABLED 0
#endif

//////////////////////////////////////////////////////////////////////////////
// System ID - conduct system identification tests on vehicle
#ifndef MODE_SYSTEMID_ENABLED
# define MODE_SYSTEMID_ENABLED HAL_LOGGING_ENABLED
#endif

//////////////////////////////////////////////////////////////////////////////
// Throw - fly vehicle after throwing it in the air
#ifndef MODE_THROW_ENABLED
# define MODE_THROW_ENABLED 1
#endif

//////////////////////////////////////////////////////////////////////////////
// ZigZag - allow vehicle to fly in a zigzag manner with predefined point A B
#ifndef MODE_ZIGZAG_ENABLED
# define MODE_ZIGZAG_ENABLED 1
#endif

//////////////////////////////////////////////////////////////////////////////
// Turtle - allow vehicle to be flipped over after a crash
#ifndef MODE_TURTLE_ENABLED
# define MODE_TURTLE_ENABLED HAL_DSHOT_ENABLED && FRAME_CONFIG != HELI_FRAME
#endif

//////////////////////////////////////////////////////////////////////////////
// Flowhold - use optical flow to hover in place
#ifndef MODE_FLOWHOLD_ENABLED
# define MODE_FLOWHOLD_ENABLED AP_OPTICALFLOW_ENABLED
#endif

//////////////////////////////////////////////////////////////////////////////

//////////////////////////////////////////////////////////////////////////////
// Weathervane - allow vehicle to yaw into wind
#ifndef WEATHERVANE_ENABLED
# define WEATHERVANE_ENABLED 1
#endif

//////////////////////////////////////////////////////////////////////////////

//////////////////////////////////////////////////////////////////////////////
// Autorotate - autonomous auto-rotation - helicopters only
#ifndef MODE_AUTOROTATE_ENABLED
#if CONFIG_HAL_BOARD == HAL_BOARD_SITL
    #if FRAME_CONFIG == HELI_FRAME
        #ifndef MODE_AUTOROTATE_ENABLED
        # define MODE_AUTOROTATE_ENABLED 1
        #endif
    #else
        # define MODE_AUTOROTATE_ENABLED 0
    #endif
#else
    # define MODE_AUTOROTATE_ENABLED 0
#endif
#endif

//////////////////////////////////////////////////////////////////////////////
// RADIO CONFIGURATION
//////////////////////////////////////////////////////////////////////////////
//////////////////////////////////////////////////////////////////////////////


//////////////////////////////////////////////////////////////////////////////
// FLIGHT_MODE
//

#ifndef FLIGHT_MODE_1
 # define FLIGHT_MODE_1                  Mode::Number::STABILIZE
#endif
#ifndef FLIGHT_MODE_2
 # define FLIGHT_MODE_2                  Mode::Number::STABILIZE
#endif
#ifndef FLIGHT_MODE_3
 # define FLIGHT_MODE_3                  Mode::Number::STABILIZE
#endif
#ifndef FLIGHT_MODE_4
 # define FLIGHT_MODE_4                  Mode::Number::STABILIZE
#endif
#ifndef FLIGHT_MODE_5
 # define FLIGHT_MODE_5                  Mode::Number::STABILIZE
#endif
#ifndef FLIGHT_MODE_6
 # define FLIGHT_MODE_6                  Mode::Number::STABILIZE
#endif


//////////////////////////////////////////////////////////////////////////////
// Throttle Failsafe
//
#ifndef FS_THR_VALUE_DEFAULT
 # define FS_THR_VALUE_DEFAULT             975
#endif

//////////////////////////////////////////////////////////////////////////////
// Takeoff
//
#ifndef PILOT_TKOFF_ALT_DEFAULT
 # define PILOT_TKOFF_ALT_DEFAULT           0     // default final alt above home for pilot initiated takeoff
#endif


//////////////////////////////////////////////////////////////////////////////
// Landing
//
#ifndef LAND_SPEED
 # define LAND_SPEED    50          // the descent speed for the final stage of landing in cm/s
#endif
#ifndef LAND_REPOSITION_DEFAULT
 # define LAND_REPOSITION_DEFAULT   1   // by default the pilot can override roll/pitch during landing
#endif
#ifndef LAND_WITH_DELAY_MS
 # define LAND_WITH_DELAY_MS        4000    // default delay (in milliseconds) when a land-with-delay is triggered during a failsafe event
#endif
#ifndef LAND_CANCEL_TRIGGER_THR
 # define LAND_CANCEL_TRIGGER_THR   700     // land is cancelled by input throttle above 700
#endif
#ifndef LAND_RANGEFINDER_MIN_ALT_CM
#define LAND_RANGEFINDER_MIN_ALT_CM 200
#endif

//////////////////////////////////////////////////////////////////////////////
// Landing Detector
//
#ifndef LAND_DETECTOR_TRIGGER_SEC
 # define LAND_DETECTOR_TRIGGER_SEC         1.0f    // number of seconds to detect a landing
#endif
#ifndef LAND_AIRMODE_DETECTOR_TRIGGER_SEC
 # define LAND_AIRMODE_DETECTOR_TRIGGER_SEC 3.0f    // number of seconds to detect a landing in air mode
#endif
#ifndef LAND_DETECTOR_MAYBE_TRIGGER_SEC
 # define LAND_DETECTOR_MAYBE_TRIGGER_SEC   0.2f    // number of seconds that means we might be landed (used to reset horizontal position targets to prevent tipping over)
#endif
#ifndef LAND_DETECTOR_ACCEL_LPF_CUTOFF
# define LAND_DETECTOR_ACCEL_LPF_CUTOFF     1.0f    // frequency cutoff of land detector accelerometer filter
#endif
#ifndef LAND_DETECTOR_ACCEL_MAX
# define LAND_DETECTOR_ACCEL_MAX            1.0f    // vehicle acceleration must be under 1m/s/s
#endif
#ifndef LAND_DETECTOR_VEL_Z_MAX
# define LAND_DETECTOR_VEL_Z_MAX              1.0f    // vehicle vertical velocity must be under 1m/s
#endif

//////////////////////////////////////////////////////////////////////////////
// Flight mode definitions
//

// Acro Mode
#ifndef ACRO_LEVEL_MAX_ANGLE
 # define ACRO_LEVEL_MAX_ANGLE      3000 // maximum lean angle in trainer mode measured in centidegrees
#endif

#ifndef ACRO_LEVEL_MAX_OVERSHOOT
 # define ACRO_LEVEL_MAX_OVERSHOOT  1000 // maximum overshoot angle in trainer mode when full roll or pitch stick is held in centidegrees
#endif

#ifndef ACRO_BALANCE_ROLL
 #define ACRO_BALANCE_ROLL          1.0f
#endif

#ifndef ACRO_BALANCE_PITCH
 #define ACRO_BALANCE_PITCH         1.0f
#endif

#ifndef ACRO_RP_EXPO_DEFAULT
 #define ACRO_RP_EXPO_DEFAULT       0.3f    // ACRO roll and pitch expo parameter default
#endif

#ifndef ACRO_Y_EXPO_DEFAULT
 #define ACRO_Y_EXPO_DEFAULT        0.0f    // ACRO yaw expo parameter default
#endif

#ifndef ACRO_THR_MID_DEFAULT
 #define ACRO_THR_MID_DEFAULT       0.0f
#endif

#ifndef ACRO_RP_RATE_DEFAULT
 #define ACRO_RP_RATE_DEFAULT      360      // ACRO roll and pitch rotation rate parameter default in deg/s
#endif

#ifndef ACRO_Y_RATE_DEFAULT
 #define ACRO_Y_RATE_DEFAULT       202.5    // ACRO yaw rotation rate parameter default in deg/s
#endif

// RTL Mode
#ifndef RTL_ALT_FINAL
 # define RTL_ALT_FINAL             0       // the altitude, in cm, the vehicle will move to as the final stage of Returning to Launch.  Set to zero to land.
#endif

#ifndef RTL_ALT
 # define RTL_ALT                   1500    // default alt to return to home in cm, 0 = Maintain current altitude
#endif

#ifndef RTL_ALT_MIN
 # define RTL_ALT_MIN               30     // min height above ground for RTL (i.e 30cm)
#endif

#ifndef RTL_CLIMB_MIN_DEFAULT
 # define RTL_CLIMB_MIN_DEFAULT     0       // vehicle will always climb this many cm as first stage of RTL
#endif

#ifndef RTL_CONE_SLOPE_DEFAULT
 # define RTL_CONE_SLOPE_DEFAULT    3.0f    // slope of RTL cone (height / distance). 0 = No cone
#endif

#ifndef RTL_MIN_CONE_SLOPE
 # define RTL_MIN_CONE_SLOPE        0.5f    // minimum slope of cone
#endif

#ifndef RTL_LOITER_TIME
 # define RTL_LOITER_TIME           5000    // Time (in milliseconds) to loiter above home before beginning final descent
#endif

// AUTO Mode
#ifndef WP_YAW_BEHAVIOR_DEFAULT
 # define WP_YAW_BEHAVIOR_DEFAULT   WP_YAW_BEHAVIOR_LOOK_AT_NEXT_WP_EXCEPT_RTL
#endif

#ifndef YAW_LOOK_AHEAD_MIN_SPEED
 # define YAW_LOOK_AHEAD_MIN_SPEED  100             // minimum ground speed in cm/s required before copter is aimed at ground course
#endif

// Super Simple mode
#ifndef SUPER_SIMPLE_RADIUS
 # define SUPER_SIMPLE_RADIUS       1000
#endif

//////////////////////////////////////////////////////////////////////////////
// Stabilize Rate Control
//
#ifndef ROLL_PITCH_YAW_INPUT_MAX
 # define ROLL_PITCH_YAW_INPUT_MAX      4500        // roll, pitch and yaw input range
#endif
#ifndef DEFAULT_ANGLE_MAX
 # define DEFAULT_ANGLE_MAX         3000            // ANGLE_MAX parameters default value
#endif

//////////////////////////////////////////////////////////////////////////////
// Stop mode defaults
//
#ifndef BRAKE_MODE_SPEED_Z
 # define BRAKE_MODE_SPEED_Z     250 // z-axis speed in cm/s in Brake Mode
#endif
#ifndef BRAKE_MODE_DECEL_RATE
 # define BRAKE_MODE_DECEL_RATE  750 // acceleration rate in cm/s/s in Brake Mode
#endif

//////////////////////////////////////////////////////////////////////////////
// PosHold parameter defaults
//
#ifndef POSHOLD_BRAKE_RATE_DEFAULT
 # define POSHOLD_BRAKE_RATE_DEFAULT    8       // default POSHOLD_BRAKE_RATE param value.  Rotation rate during braking in deg/sec
#endif
#ifndef POSHOLD_BRAKE_ANGLE_DEFAULT
 # define POSHOLD_BRAKE_ANGLE_DEFAULT   3000    // default POSHOLD_BRAKE_ANGLE param value.  Max lean angle during braking in centi-degrees
#endif

//////////////////////////////////////////////////////////////////////////////
// Pilot control defaults
//

#ifndef THR_DZ_DEFAULT
# define THR_DZ_DEFAULT         100             // the deadzone above and below mid throttle while in althold or loiter
#endif

// default maximum vertical velocity and acceleration the pilot may request
#ifndef PILOT_VELZ_MAX
 # define PILOT_VELZ_MAX    250     // maximum vertical velocity in cm/s
#endif
#ifndef PILOT_ACCEL_Z_DEFAULT
 # define PILOT_ACCEL_Z_DEFAULT 250 // vertical acceleration in cm/s/s while altitude is under pilot control
#endif

#ifndef PILOT_Y_RATE_DEFAULT
 # define PILOT_Y_RATE_DEFAULT  202.5   // yaw rotation rate parameter default in deg/s for all mode except ACRO
#endif
#ifndef PILOT_Y_EXPO_DEFAULT
 # define PILOT_Y_EXPO_DEFAULT  0.0     // yaw expo parameter default for all mode except ACRO
#endif

#ifndef AUTO_DISARMING_DELAY
# define AUTO_DISARMING_DELAY  10
#endif

//////////////////////////////////////////////////////////////////////////////
// Throw mode configuration
//
#ifndef THROW_HIGH_SPEED
# define THROW_HIGH_SPEED       500.0f  // vehicle much reach this total 3D speed in cm/s (or be free falling)
#endif
#ifndef THROW_VERTICAL_SPEED
# define THROW_VERTICAL_SPEED   50.0f   // motors start when vehicle reaches this total 3D speed in cm/s
#endif

//////////////////////////////////////////////////////////////////////////////
// Logging control
//

// Default logging bitmask
#ifndef DEFAULT_LOG_BITMASK
 # define DEFAULT_LOG_BITMASK \
    MASK_LOG_ATTITUDE_MED | \
    MASK_LOG_GPS | \
    MASK_LOG_PM | \
    MASK_LOG_CTUN | \
    MASK_LOG_NTUN | \
    MASK_LOG_RCIN | \
    MASK_LOG_IMU | \
    MASK_LOG_CMD | \
    MASK_LOG_CURRENT | \
    MASK_LOG_RCOUT | \
    MASK_LOG_OPTFLOW | \
    MASK_LOG_PID | \
    MASK_LOG_COMPASS | \
    MASK_LOG_CAMERA | \
    MASK_LOG_MOTBATT
#endif

//////////////////////////////////////////////////////////////////////////////
// Fence, Rally and Terrain and AC_Avoidance defaults
//

#if MODE_FOLLOW_ENABLED && !AP_AVOIDANCE_ENABLED
  #error Follow Mode relies on AP_AVOIDANCE_ENABLED which is disabled
#endif

#if MODE_AUTO_ENABLED && !MODE_GUIDED_ENABLED
  #error ModeAuto requires ModeGuided which is disabled
#endif

#if MODE_AUTO_ENABLED && !MODE_CIRCLE_ENABLED
  #error ModeAuto requires ModeCircle which is disabled
#endif

#if MODE_AUTO_ENABLED && !MODE_RTL_ENABLED
  #error ModeAuto requires ModeRTL which is disabled
#endif

#if FRAME_CONFIG == HELI_FRAME && !MODE_ACRO_ENABLED
  #error Helicopter frame requires acro mode support which is disabled
#endif

#if MODE_SMARTRTL_ENABLED && !MODE_RTL_ENABLED
  #error SmartRTL requires ModeRTL which is disabled
#endif

#if HAL_ADSB_ENABLED && !MODE_GUIDED_ENABLED
  #error ADSB requires ModeGuided which is disabled
#endif

#if MODE_FOLLOW_ENABLED && !MODE_GUIDED_ENABLED
  #error Follow requires ModeGuided which is disabled
#endif

#if MODE_GUIDED_NOGPS_ENABLED && !MODE_GUIDED_ENABLED
  #error ModeGuided-NoGPS requires ModeGuided which is disabled
#endif

//////////////////////////////////////////////////////////////////////////////
// Developer Items
//

#ifndef AP_COPTER_ADVANCED_FAILSAFE_ENABLED
# define AP_COPTER_ADVANCED_FAILSAFE_ENABLED 0
#endif

#ifndef CH_MODE_DEFAULT
 # define CH_MODE_DEFAULT   5
#endif

#ifndef TOY_MODE_ENABLED
#define TOY_MODE_ENABLED 0
#endif

#if TOY_MODE_ENABLED && FRAME_CONFIG == HELI_FRAME
  #error Toy mode is not available on Helicopters
#endif

#define HAL_FRAME_TYPE_DEFAULT AP_Motors::MOTOR_FRAME_TYPE_X // LOCKED: Hex X

#ifndef AC_CUSTOMCONTROL_MULTI_ENABLED
#define AC_CUSTOMCONTROL_MULTI_ENABLED FRAME_CONFIG == MULTICOPTER_FRAME && AP_CUSTOMCONTROL_ENABLED
#endif

#ifndef AC_PAYLOAD_PLACE_ENABLED
#define AC_PAYLOAD_PLACE_ENABLED 1
#endif

#ifndef USER_PARAMS_ENABLED
  #define USER_PARAMS_ENABLED 0
#endif
